tokio-boring/src/lib.rs

```diff
@@ -307,6 +307,40 @@ impl<S> HandshakeError<S> {
             _ => None,
         }
     }
+
+    /// Returns a reference to the inner SSL error, if this error was caused by
+    /// a SSL error.
+    pub fn as_ssl_error(&self) -> Option<&boring::error::ErrorStack> {
+        match &self.0 {
+            ssl::HandshakeError::Failure(s) => s.error().ssl_error(),
+            ssl::HandshakeError::SetupFailure(ref s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// Consumes `self` and returns the inner I/O error by value, if this error
+    /// was caused by an I/O error.
+    pub fn into_io_error(self) -> Option<io::Error> {
```

Comment:
FWIW I'm not sure that this is valuable for us... If it's not an Io error we lose the original error?

I guess we'd have to use something like `if e.as_io_error().is_some() { e.into_io_error.unwrap() {` ? that's pretty cumbersome

Reply:
I thought about adding an `is_io`/`is_ssl`, but I wasn't sure if it was necessary given that it would just be `as_io_error().is_some()`. I could add that as well, though. IMO it's still less cumbersome than the current code we have, which calls `as_io_error` and creates a new `io::Error` using values from the one that's passed by reference.

Unfortunately, we can't

Reply:
@olix0r okay, I've modified these to return `Result<io::Error, Self>` and `Result<ErrorStack, Self>` so you can now write code like
```rust
error
    .into_ssl_error()
    .map(Into::into)
    .or_else(|error| error.into_io_error().map(Into::into))
```
or similar

Reply:
awesome!